RecoLocalCalo/HcalRecProducers/plugins/alpaka/Mahi.dev.cc

```diff
@@ -0,0 +1,1481 @@
+#include <iostream>
+#include <alpaka/alpaka.hpp>
+#include <Eigen/Dense>
+
+#include "HeterogeneousCore/AlpakaInterface/interface/workdivision.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/atomicMaxPair.h"
+
+#include "DataFormats/CaloRecHit/interface/MultifitComputations.h"
+// needed to compile with USER_CXXFLAGS="-DCOMPUTE_TDC_TIME"
+#include "DataFormats/HcalRecHit/interface/HcalSpecialTimes.h"
+#include "FWCore/Utilities/interface/CMSUnrollLoop.h"
+// TODO reuse some of the HCAL constats from
+//#include "RecoLocalCalo/HcalRecAlgos/interface/HcalConstants.h"
+
+#include "Mahi.h"
+
+#ifdef HCAL_MAHI_GPUDEBUG
+#define DETID_TO_DEBUG 1125647428
+#endif
+
+namespace ALPAKA_ACCELERATOR_NAMESPACE {
+
+  using namespace cms::alpakatools;
+  namespace hcal::reconstruction {
+    namespace mahi {
+
+      ALPAKA_FN_HOST_ACC ALPAKA_FN_INLINE float uint_as_float(unsigned long long int val) {
+#if defined(__CUDA_ARCH__) or defined(__HIP_DEVICE_COMPILE__)
+        return __uint_as_float(val);
+#else
+        return edm::bit_cast<float>(static_cast<unsigned int>(val));
+#endif
```

Comment:
(likely beyond this PR)

I wonder how we should try to improve this. In a way the added `atomicMaxPair()` has an abstraction problem, because it explicitly uses `unsigned long long*` for the atomic communication across threads, and packing of the argument `std::pair<unsigned, float>` into the `unsigned long long` is an implementation detail of `atomicMaxPair()`. But then the calling code needs to unpack the `unsigned long long` into `unsigned, float`.

I started to experiment with a specific `PackedPair64` class template that would provide `uint64_t*` accessor to be used with the `atomicCAS()`, and `first()` and `second()` accessors to unpack the pair elements. That would keep the packing and unpacking as implementation details for all use cases. The `#ifdef` on `__CUDA_ARCH__` and `__HIP_DEVICE_COMPILE__` are still there. I'd hope the C++20 `std::bit_cast` would be usable in the (CUDA and ROCm) device code, but I don't know.

Reply:
I think the `bit_cast` part we can wrap and implement inside Alpaka, something like `alpaka::bit_cast<T>(acc, value)`.

C++20 `std::bit_cast` is `constexpr`, so technically usable in CUDA and ROCm code, but I do not know if it will do the right thing there.